File: src/client/controller/LoginController.java
package client.controller;

import client.model.ClientApplicationModel;
import client.model.LoginModel;
import client.view.ClientApplicationView;
import client.view.LoginView;
import shared.SwingResources;

import javax.swing.*;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.sql.SQLException;
import java.util.Arrays;

/**
 * The LoginController processes user requests when trying to access the application through their existing account.
 */
public class LoginController {
    /**
     * The view.
     */
    private LoginView view;
    /**
     * The model.
     */
    private LoginModel model;

    /**
     * Constructs a LoginController with a specified view and model.
     * @param view The specified view.
     * @param model The specified model.
     */
    public LoginController(LoginView view, LoginModel model) {
        this.view = view;
        this.model = model;

        // action listeners
        view.setLoginListener(new LoginListener());
        view.setShowPasswordListener(new SwingResources.ShowPasswordListener(view.getChkShowPassword(),
                view.getTxtPassword()));

        // mouse listeners
        view.getBtnLogin().addMouseListener(new SwingResources.CursorChanger(view.getBtnLogin()));

        // focus listeners
        view.getTxtUsername().addFocusListener(new SwingResources.TextFieldFocus(view.getTxtUsername(),
                "Username", view.getLblErrorMessage()));
        view.getTxtPassword().addFocusListener(new SwingResources.PasswordFocusWithCheckbox(view.getTxtPassword(),
                view.getChkShowPassword(), "Password", view.getLblErrorMessage()));

        view.revalidate();
        view.repaint();
    }

    /**
     * Processes the request for loggin in.
     *
     * FOR CODE CLEANING
     */
    class LoginListener implements ActionListener {
        /**
         * T
         * @param e The specified action event.
         */
        @Override
        public void actionPerformed(ActionEvent e) {
            String username = view.getTxtUsername().getText();
            try {
                if (model.validateAccount(view.getTxtUsername().getText(), Arrays.toString(view.getTxtPassword().getPassword()))){
                    SwingUtilities.invokeLater(() -> new ClientApplicationController(new ClientApplicationView(),
                            new ClientApplicationModel(username, model.getWfImpl())));
                    view.dispose();
                } else {
                    SwingUtilities.invokeLater(() -> {
                        view.setErrorMessage("Wrong credentials. Try again.");
                        view.getTxtPassword().setText("Password");
                        view.getTxtPassword().setEchoChar((char) 0);
                        view.getChkShowPassword().setSelected(false);
                    });
                }
            } catch (SQLException sqle){
                sqle.printStackTrace();
            } catch (Exception ex) {
                throw new RuntimeException(ex);
            }
        }
    }
}

File: src/Client_Python/yuen_client/Controller/LoginController.py
'''
import javax.swing as swing
import java.awt.event as event
import shared.SwingResources as swing_resources
from client.model import ClientApplicationModel, LoginModel
from client.view import ClientApplicationView, LoginView

class LoginController:
    def __init__(self, view, model):
        self.view = view
        self.model = model

        # action listeners
        self.view.set_password_listener(LoginListener())
        self.view.set_login_listener(LoginListener())
        self.view.set_show_password_listener(swing_resources.ShowPasswordListener(self.view.get_chk_show_password(),
                                                                                  self.view.get_txt_password()))

        # mouse listeners
        self.view.get_btn_login().add_mouse_listener(swing_resources.CursorChanger(self.view.get_btn_login()))

        # focus listeners
        self.view.get_txt_username().add_focus_listener(swing_resources.TextFieldFocus(self.view.get_txt_username(),
                                                                                        "Username",
                                                                                        self.view.get_lbl_error_message()))
        self.view.get_txt_password().add_focus_listener(swing_resources.PasswordFocusWithCheckbox(self.view.get_txt_password(),
                                                                                                   self.view.get_chk_show_password(),
                                                                                                   "Password",
                                                                                                   self.view.get_lbl_error_message()))

        self.view.revalidate()
        self.view.repaint()

class LoginListener(event.ActionListener):
    def actionPerformed(self, e):
        username = self.view.get_txt_username().getText()
        try:
            msg = self.model.validate_account(self.view.get_txt_username().getText(),
                                              self.view.get_txt_password().getText())
            if msg == "valid":
                swing.SwingUtilities.invokeLater(lambda: ClientApplicationController(ClientApplicationView(),
                                                                                     ClientApplicationModel(username,
                                                                                                            self.model.get_wf_impl())))
                self.view.dispose()
            else:
                swing.SwingUtilities.invokeLater(lambda: (
                    self.view.set_error_message(msg),
                    self.view.get_txt_password().setText("Password"),
                    self.view.get_txt_password().setEchoChar(chr(0)),
                    self.view.get_chk_show_password().setSelected(False)
                ))
        except Exception as ex:
            raise RuntimeException(ex)
'''